source/client.c

```diff
@@ -125,6 +125,7 @@ static void s_mqtt_client_shutdown(
             case AWS_MQTT_CLIENT_STATE_DISCONNECTING:
                 /* disconnect requested by user */
                 /* Successfully shutdown, so clear the outstanding requests */
+                /* TODO: this also invokes user callback, which is extremely dangerous */
```

Comment:
Honestly I think we should ban hash tables that do anything other than dead-simple deallocation of their key-value members in these callbacks.

Reply:
yeah, agreed. I'll try to improve this.

Reply:
removed from the other PR, https://github.com/awslabs/aws-c-mqtt/pull/154/commits/22272ebc962874844d24e049484d7297269117d7